### PR TITLE
wardeploy improvements to support custom app names

### PR DESCRIPTION
**Pull request by shrishrirang**
- This change allows deploying war files to web apps other than ROOT.
- Posting war file contents to /api/wardeploy?name=myappname will create the myappname webapp by deploying the war contents to wwwroot/webapps/<appname> directory. If no app name is specified, ROOT is inferred.

### CODE REVIEW

**Comment by shrishrirang** [comment]
@davidebbo again as Nick is oof. This PR uses https://github.com/projectkudu/KuduSync/pull/34 and https://github.com/projectkudu/KuduScript/pull/94 to perform clean deployment by ignoring the manifest when performing wardeploy. It does this by setting `CleanupTargetDirectory = true`.

File: Kudu.Core/Deployment/DeploymentManager.cs [hunk 1]
```diff
@@ -624,6 +624,7 @@ private DeployResult GetResult(string id, string activeDeploymentId, bool isDepl
                 {
                     NextManifestFilePath = GetDeploymentManifestPath(id),
                     PreviousManifestFilePath = GetActiveDeploymentManifestPath(),
+                    IgnoreManifest = deploymentInfo.CleanupTargetDirectory,
```
Comment (davidebbo) [inline Kudu.Core/Deployment/DeploymentManager.cs]: It's not instantly clear to me how ignoring the manifest results in cleaning up the target folder. A comment would be useful here.

Comment (shrishrirang) [reply Kudu.Core/Deployment/DeploymentManager.cs]: Done.

Comment (davidebbo) [reply Kudu.Core/Deployment/DeploymentManager.cs]: Thanks, I guess I had misunderstood/forgotten when `IgnoreManifest` meant, but now I understand.

**Comment by shrishrirang** [comment]
Thanks for the review!

**Comment by davidebbo** [comment]
@shrishrirang darn, it looks like this completely breaks Kudu git deployments on Windows. I haven't looked at why yet.

File: Kudu.Core/Deployment/DeploymentManager.cs [hunk 2]
```diff
@@ -624,6 +624,11 @@ private DeployResult GetResult(string id, string activeDeploymentId, bool isDepl
                 {
                     NextManifestFilePath = GetDeploymentManifestPath(id),
                     PreviousManifestFilePath = GetActiveDeploymentManifestPath(),
+                    IgnoreManifest = deploymentInfo.CleanupTargetDirectory, // Ignoring the manifest will cause kudusync to delete sub-directories / files
```
Comment (davidebbo) [inline Kudu.Core/Deployment/DeploymentManager.cs]: Problem is that `deploymentInfo` can be null, which crashes here.

**Comment by shrishrirang** [comment]
Aah! I'll share a PR for the fix.

**Comment by davidebbo** [comment]
I just pushed a fix to just check for null. 79c21de2a24c8425bbb00c1103a17f2ddbbc0a37.

**Comment by shrishrirang** [comment]
Thanks! :)

**Comment by davidebbo** [comment]
Now the only test that failed is TestSimpleWarDeploymentPerformsCleanDeployment. Is that supposed to run at all on Windows?

**Comment by shrishrirang** [comment]
Yes, it is supposed to run successfully. That tests tests the wardeploy performs a clean deployment, which is one of the changes introduced by this PR. It did work fine on my machine. Let me take a look.

**Comment by davidebbo** [comment]
Seems the file-before-wardeployment doesn't get deleted.

**Comment by shrishrirang** [comment]
Ok, let me check why.

**Comment by davidebbo** [comment]
Doesn't it need the new kuduscript that you sent a PR for? https://github.com/projectkudu/KuduScript/pull/94

**Comment by shrishrirang** [comment]
Oh yes, that is precisely the reason!
In my mind everything was supposed to work fine because the `IGNORE_MANIFEST` environment variable would have no effect until kuduscript changes are in. Didn't realize I already added a test and that would fail. Fix for this is to just comment the test for now. Once I update `updateNodeModules.cmd`, it needs to be enabled again. Should I push that change or you have it already ready to push? :)

**Comment by davidebbo** [comment]
It's ok if this one test fails for a day or two.

So in what scenario exactly will IgnoreManifest be set? Only the WAR scenario on Linux? Because if it's more general, then technically all the generated scripts should support it, as they all call kudusync.

**Pull request merged**